extensions/interactions/AlgebraicExpressionInput/directives/algebraic-expression-input-validation.service.ts

```diff
@@ -0,0 +1,110 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Validator service for the AlgebraicExpressionInput interaction.
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+import { AnswerGroup } from
+  'domain/exploration/AnswerGroupObjectFactory';
+import { IWarning, baseInteractionValidationService } from
+  'interactions/base-interaction-validation.service';
+import { IAlgebraicExpressionInputCustomizationArgs } from
+  'extensions/interactions/customization-args-defs';
+import { AlgebraicExpressionInputRulesService } from
+  './algebraic-expression-input-rules.service';
+import { Outcome } from
+  'domain/exploration/OutcomeObjectFactory';
+import { AppConstants } from 'app.constants';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class AlgebraicExpressionInputValidationService {
+  constructor(
+      private baseInteractionValidationServiceInstance:
+        baseInteractionValidationService) {}
+
+  getAllWarnings(
+      stateName: string,
+      customizationArgs: IAlgebraicExpressionInputCustomizationArgs,
+      answerGroups: AnswerGroup[], defaultOutcome: Outcome): IWarning[] {
+    var warningsList = [];
+    var aeirs = (
+      new AlgebraicExpressionInputRulesService());
+
+    warningsList = warningsList.concat(
+      this.baseInteractionValidationServiceInstance.getAllOutcomeWarnings(
+        answerGroups, defaultOutcome, stateName));
+
+
+    // This validations ensures that there are no redundant rules present in the
+    // answer groups.
+    // An IsEquivalentTo rule will make all of the following rules with a
+    // matching input, invalid.
+    // A MatchesExactlyWith rule will make the following rules of the same rule
+    // type and a matching input, invalid.
+    var seenRules = [];
+
+    for (var i = 0; i < answerGroups.length; i++) {
+      var rules = answerGroups[i].rules;
+      for (var j = 0; j < rules.length; j++) {
+        var currentInput = <string> rules[j].inputs.x;
+        var currentRuleType = <string> rules[j].type;
+
+        for (var seenRule of seenRules) {
+          var seenInput = <string> seenRule.inputs.x;
+          var seenRuleType = <string> seenRule.type;
+
+          if (seenRuleType === 'IsEquivalentTo') {
+            // This rule will make all of the following matching
+            // inputs obsolete.
+            if (aeirs.IsEquivalentTo(seenInput, {x: currentInput})) {
```

Comment:
Merge this with the above condition using AND?

Reply:
Done.